### ISSUE TITLE
Move HTTP-specific things from runtime to links

### ISSUE THREAD

**Issue by KATT** [trpc/trpc#2709]
tRPC isn't tied to HTTP so we shouldn't actually enforce using HTTP when creating the TRPC client.

This could be done by deprecating those fields like the below, but then we have no _guarantee_ when creating the HTTPLinks that you do it right.

```diff
interface CreateTRPCClientBaseOptions {
  /**
   * Add ponyfill for fetch
+   * @deprecated moved to HTTP Links
   */
  fetch?: typeof fetch;
  /**
+   * @deprecated moved to HTTP Links
   * Add ponyfill for AbortController
   */
  AbortController?: typeof AbortController;
  /**
   * Headers to be set on outgoing requests or a callback that of said headers
   * @link http://localhost:3000/docs/v10/header
+   * @deprecated moved to HTTP Links
   */
  headers?: HTTPHeaders | (() => HTTPHeaders | Promise<HTTPHeaders>);
  /**
   * Data transformer
   * @link https://trpc.io/docs/data-transformers
   **/
  transformer?: ClientDataTransformerOptions;
}
```

And

```diff

export interface HTTPLinkOptions {
  url: string;
+  fetch?: TRPCFetch;
+  AbortController?: typeof AbortController;
+  headers?: () => HTTPHeaders | Promise<HTTPHeaders>;
}
```

_Originally posted by @KATT in https://github.com/trpc/trpc/issues/2700#issuecomment-1245930214_

**Comment by moltar** [comment]
I also want to suggest defining a narrow TypeScript interface that is a partial of the `fetch` default interface and using that instead of `typeof fetch`. That way, hopefully, other `fetch` implementations can fit the interface.

I tried both `node-fetch` and `undici.fetch`, and both throw typing errors.

This interface should only define types that are actually used by the tRPC internals and not the entire kitchen sink that fetch has.

**Comment by moltar** [comment]
> but then we have no guarantee when creating the HTTPLinks that you do it right.

I think maybe a new package can be created that is HTTP-specific: `@trpc/client-http` that has sane defaults.

Or a new factory function `createHttpClient(...)`.

**Comment by KATT** [comment]
Addressed in #2723, added you as a contributor @moltar - would love your review on it :)
